src/index.ts

```diff
@@ -0,0 +1,49 @@
+// NOTE: We use discriminated unions over "success" property which simplifies the usage
+export type GoResultSuccessArray<T> = [null, T];
+export type GoResultSuccessObject<T> = { data: T; success: true };
+export type GoResultSuccess<T> = GoResultSuccessArray<T> & GoResultSuccessObject<T>;
+
+export type GoResultErrorArray<E extends Error = Error> = [E, null];
+export type GoResultErrorObject<E extends Error = Error> = { error: E; success: false };
+export type GoResultError<E extends Error = Error> = [E, null] & { error: E; success: false };
+
+export type GoResult<T, E extends Error = Error> = GoResultSuccess<T> | GoResultError<E>;
+
+export const success = <T>(value: T): GoResultSuccess<T> => {
+  const result: any = [null, value];
+  result.data = value;
+  return result;
+};
+
+export const fail = <E extends Error>(err: Error): GoResultError<E> => {
+  const result: any = [err, null];
+  result.error = err;
+  return result;
+};
```

Comment:
Even though the typing says that there's a `success` property available it's not true, you're not setting it.

Reply:
Thanks! I typed the result as `any` but then TS didn't warn me about the missing part. I'll fix this.